warehouse/oidc/models/gitlab.py

```diff
@@ -0,0 +1,288 @@
+    # where REF is the value of the `ref_path` claim.
+
+    # Defensive: GitLab should never give us an empty ci_config_ref_uri,
+    # but we check for one anyway just in case.
+    if not signed_claim:
+        raise InvalidPublisherError("The ci_config_ref_uri claim is empty")
+
+    # We need at least one of these to be non-empty
+    ref = all_signed_claims.get("ref_path")
+    sha = all_signed_claims.get("sha")
+    if not (ref or sha):
+        raise InvalidPublisherError("The ref_path and sha claims are empty")
+
+    expected = {f"{ground_truth}@{_ref}" for _ref in [ref, sha] if _ref}
+    if signed_claim not in expected:
+        raise InvalidPublisherError(
+            "The ci_config_ref_uri claim does not match, expecting one of "
+            f"{sorted(expected)!r}, got {signed_claim!r}"
+        )
+
+    return True
+
+
+def _check_environment(ground_truth, signed_claim, all_signed_claims):
+    # When there is an environment, we expect a string.
+    # For tokens that are generated outside of an environment, the claim will
+    # be missing.
+
+    # If we haven't set an environment name for the publisher, we don't need to
+    # check this claim
+    if ground_truth == "":
+        return True
+
+    # Defensive: GitLab might give us an empty environment if this token wasn't
+    # generated from within an environment, in which case the check should
+    # fail.
+    if not signed_claim:
+        return False
+
+    return ground_truth == signed_claim
+
+
+def _check_sub(ground_truth, signed_claim, _all_signed_claims):
+    # We expect a string formatted as follows:
+    # project_path:NAMESPACE/PROJECT[:OPTIONAL-STUFF]
+    # where :OPTIONAL-STUFF is a concatenation of other job context
+    # metadata. We currently lack the ground context to verify that
+    # additional metadata, so we limit our verification to just the
+    # NAMESPACE/PROJECT component.
+
+    # Defensive: GitLab should never give us an empty subject.
+    if not signed_claim:
+        return False
+
+    components = signed_claim.split(":")
+    if len(components) < 2:
+        return False
+
+    namespace, project, *_ = components
+    if not namespace or not project:
+        return False
+
+    # The sub claim is case-insensitive
+    return f"{namespace}:{project}".lower() == ground_truth.lower()
+
+
+class GitLabPublisherMixin:
+    """
+    Common functionality for both pending and concrete GitLab OIDC publishers.
+    """
+
+    namespace = mapped_column(String, nullable=False)
+    project = mapped_column(String, nullable=False)
+    workflow_filepath = mapped_column(String, nullable=False)
+    environment = mapped_column(String, nullable=False)
+
+    __required_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        "sub": _check_sub,
+        "project_path": check_claim_binary(str.__eq__),
+        "ci_config_ref_uri": _check_ci_config_ref_uri,
+    }
+
+    __required_unverifiable_claims__: set[str] = {"ref_path", "sha"}
```

Comment:
This has the effect of requiring both claims to be present and not `None`: https://github.com/pypi/warehouse/blob/fd35a53dfc4b2aae994aa9368908097ec5b15552/warehouse/oidc/models/_core.py#L189-L204

But based on _check_ci_config_ref_uri, it seems like at least one of these could be non-empty? We might need to special-case the validation here for these.

Reply:
This was due to using the Github code as a base, which assumes either `sha` or `ref_path` can be empty. GitLab's documentation specifies that both should be present, so I changed `_check_ci_config_ref_uri` to reflect that.